Source/Core/Common/HttpRequest.cpp

```diff
@@ -0,0 +1,119 @@
+// Copyright 2017 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include "Common/HttpRequest.h"
+
+#include <cstddef>
+#include <curl/curl.h>
+
+#include "Common/Logging/Log.h"
+
+namespace Common
+{
+class HttpRequest::Impl final
+{
+public:
+  enum class Method
+  {
+    GET,
+    POST,
+  };
+
+  Impl();
+
+  bool IsValid() const;
+  Response Fetch(const std::string& url, Method method, const u8* payload, size_t size);
+
+private:
+  std::unique_ptr<CURL, decltype(&curl_easy_cleanup)> m_curl{curl_easy_init(), curl_easy_cleanup};
+};
+
+HttpRequest::HttpRequest() : m_impl(std::make_unique<Impl>())
+{
+}
+
+HttpRequest::~HttpRequest() = default;
+
+bool HttpRequest::IsValid() const
+{
+  return m_impl->IsValid();
+}
+
+HttpRequest::Response HttpRequest::Get(const std::string& url)
+{
+  return m_impl->Fetch(url, Impl::Method::GET, nullptr, 0);
+}
+
+HttpRequest::Response HttpRequest::Post(const std::string& url, const std::vector<u8>& payload)
+{
+  return m_impl->Fetch(url, Impl::Method::POST, payload.data(), payload.size());
+}
+
+HttpRequest::Response HttpRequest::Post(const std::string& url, const std::string& payload)
+{
+  return m_impl->Fetch(url, Impl::Method::POST, reinterpret_cast<const u8*>(payload.data()),
+                       payload.size());
+}
+
+HttpRequest::Impl::Impl()
+{
+  if (!m_curl)
+    return;
+
+  // libcurl may not have been built with async DNS support, so we disable
+  // signal handlers to avoid a possible and likely crash if a resolve times out.
+  curl_easy_setopt(m_curl.get(), CURLOPT_NOSIGNAL, true);
+  curl_easy_setopt(m_curl.get(), CURLOPT_TIMEOUT, 3);
+#ifdef _WIN32
+  // ALPN support is enabled by default but requires Windows >= 8.1.
+  curl_easy_setopt(m_curl.get(), CURLOPT_SSL_ENABLE_ALPN, false);
+#endif
+}
+
+bool HttpRequest::Impl::IsValid() const
+{
+  return m_curl != nullptr;
+}
+
+static size_t CurlCallback(u8* data, size_t size, size_t nmemb, std::vector<u8>* buffer)
```

Comment:
Missed this initially, but [CURLOPT_WRITEFUNCTION](https://curl.haxx.se/libcurl/c/CURLOPT_WRITEFUNCTION.html) callbacks need to have a signature like the following:

```cpp
size_t write_callback(char *ptr, size_t size, size_t nmemb, void *userdata);
```

So the `u8*` and `std::vector<u8>*` pointers need to be changed. The void pointer will need to be `static_cast`ed to the correct type.